---
title: "CircaPower"
output: rmarkdown::html_vignette
bibliography: ref.bib
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{CircaPower}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, eval=FALSE}
library(CircaPower)
```


# Introduction

## Background

Circadian clocks are 24-hour endogenous oscillators in physiological and behavioral processes. Though recent transcriptomic studies have been successful in revealing the circadian rhythmicity in gene expression, the power calculation and study design for omics circadian analysis have not been explored. CircaPower is an analytical framework for statistical power calculation in circadian gene detection analysis. It is based on the cosinor model, which is the only rhythmicity detection model with explicitly defined effect size and data variability and allows for irregular Zeitgeber time (ZT; standardized diurnal time with ZT0 for the beginning of day and ZT12 for the beginning of night) distribution. CircaPower is designed for fast and accurate power caculation for both passive and active sampling design. In passive design, investigators have no control of the collected ZT. Such a passive design is commonly seen in studies with human tissues that are difficult to obtain e.g., post-mortem brain tissues (@chen2016effects, @li2013circadian, @seney2019diurnal). In contrast, investigators have full control of the sample collection time in an active sampling design. Such an active design is commonly seen in animal studies (@zhang2014circadian)or human blood studies (@moller2013effects).


## Statistical method
The sinusoidal wave curve assumed in the cosinor model (@cornelissen2014cosinor) is illustrated in Figure 1, where y is the expression value for a gene, t is the ZT, M is the MESOR (Midline Estimating Statistic Of Rhythm, a rhythm-adjusted mean), A is the amplitude and $\phi$ is the phase shift. $\omega$ is the frequency of the sinusoidal wave i.e, $\omega = \frac{2\pi}{Period}$. Without loss of generality, we set $period = 24$ hours to mimic the diurnal period. \\

For a given sample $i$ ($1\le i \le n$, $n$ is the total number of samples), denote by $y_i$ the expression value of a gene and 
$t_i$ the observed ZT. 
We assume sinusoidal wave function:

\begin{equation}
\label{eq:sin1}
y_i=A\sin(\omega(t_i+\phi))+M+\varepsilon_i, 
\end{equation}

\noindent where $\varepsilon_i$ is the error term for sample $i$; 
we assume $\varepsilon_i$'s are identically and independently distributed ($i.i.d.$) from $\varepsilon_i \sim  \textit{N}(0, \sigma^2)$, 
where $\sigma$ is the noise level. 

```{r ,echo = FALSE,out.width='50%',fig.align="center",fig.cap="Figure 1: The sinusoidal wave curve underlying circadian rhythmicity power calculation framework."}
knitr::include_graphics("circadianCurve.png")
```

Equivalently, we could re-write the model as 

\begin{equation}
\label{eq:sin2}
y_i=E\sin(\omega t_i) + H\cos(\omega t_i) +M+\varepsilon_i, 
\end{equation}

\noindent where $E = A \cos (\omega \phi)$, and $H = A \sin (\omega \phi)$. It turns into a linear regression problem with the corresponding F statistics as $F^{stat} = \frac{\frac{TSS - RSS}{r - 1}}{\frac{RSS}{n-r}}$ where where $RSS = \sum^n_{i=1} (y_i-\hat{y_i})^2$, $TSS = \sum^n_{i=1} (y_i-\bar{y})^2$, $\hat{y_i} = \hat{A} \sin(\omega(t_i+\hat{\phi})) + \hat{M}$, $\bar{y} = \sum_i y_i / n$, with $\hat{A}$, $\hat{\phi}$, and  $\hat{M}$ estimated under least squared loss.\

\

From the linear model theory, under the null hypothesis that there is no circadian rhythmicity, i.e., $A=0$, or equivalently, $E=H=0$,
$$F^{stat} \sim f_0(\cdot|2, n - 3),$$
where $2$ and $n-3$ are the degrees of freedom of the $F$ distribution, and $f_0$ denotes a regular $F$ distribution with non-centrality parameter $0$.\

\

Under the alternative hypothesis that there exists circadian rhythmicity pattern, i.e., $A \ne 0$, or equivalently, $E\ne 0$ or $H\ne 0$, 
$$F^{stat} \sim  f_\lambda(\cdot|2,n-3), \lambda  = \frac{A^2 }{\sigma^2} \sum_i \sin^2(w(t_i + \phi))$$
where $2$ and $n-3$ are the degrees of freedom of the $F$ distribution, and $f_\lambda$ denotes a non-central $F$ distribution with non-centrality parameter $\lambda$.\

\

Figure 2 shows the relationship between the null and alternative distributions. By assuming the rejection boundary is at alpha level $\alpha$, the relationship between $\alpha$ and the power $1-\beta$ is:
$$F_{\lambda}^{-1} (\beta| 2 , n - 3) = F_0^{-1} (1 - \alpha| 2 , n - 3)$$
\noindent where $F_\lambda (x| df_1 , df_2)$ and $F_0 (x| df_1 , df_2)$ are the cumulative density function of $f_\lambda (\cdot|df_1 , df_2)$ and $f_0 (\cdot|df_1 , df_2)$ evaluated at $x$ respectively. Therefore, the power $1=\beta$ can be derived from the equation above.


```{r,echo = FALSE,out.width='50%',fig.align="center",fig.cap="Figure 2: the relationship between power and type I control in detecting circadian rhythmicity. Black curve: the density function of the F statistics under the null distribution; Blue curve: the density function of the F statistics under the alternative distribution; Red dashed line: the decision boundary (i.e., $F^*$)."}
knitr::include_graphics("power_Ftest.png")
```

Under fixed type I error control $\alpha$, the detection power depends on the total effect size $\lambda$ which can be decomposed into three factors: (1) sample size $n$; (2) intrinsic effect size $r=A/\sigma$; (3) sampling design factor $d = \frac{1}{n}\sum_{i=1}^n \sin^2(w(t_i + \phi))$. In our paper, we show that for active design (i.e. evenly spaced ZT distribuiton), the sampling design factor $d$ is a constant of $\frac{1}{2}$ independent of the $\phi$ as long as the number of sampling times is greater or equal to 3 (i.e., phase-invariant property).

## About this tutorial

This is a tutorial for the usage of the CircaPower package.
The major contents of this tutorial includes:

- Circadian power calculation (active design) when the circadian times are evenly spaced.
- Circadian power calculation (passive design) given irregular circadian time distribuiton.
- A case study how to perform circadian power calculation using mouse skeletal muscle pilot data.
- Intrinsic effect size from public databases.


# About the package

## How to install the package

To install this package, start R (version "3.0" or above) and enter:

```{r, eval=FALSE}
library(devtools)
install_github("https://github.com/circaPower/CircaPower") 
```

## How to cite the package


- Wei Zong, Marianne L. Seney, Kyle D. Ketchesin, Michael T. Gorczyca, Andrew C. Liu, Karyn A. Esser, George C. Tseng*, Colleen A. McClung* and Zhiguang Huo*. Experimental Design and Power Calculation in Omics Circadian Rhythmicity Detection. (Submitted, *: co-coresponding authors)

- The manuscript can be found here: to be updated

## Maintainer

Wei Zong (wez97@pitt.edu) and Zhiguang Huo (zhuo@ufl.edu)


# Examples

## 1. Circadian power calculation (active design) when the circadian times are evenly spaced.

Because of the phase invariant property of evenly-spaced sampling design, the detection power only depends on the total number of samples n and intrinsic effect size r given type I error control level $\alpha$.  
```{r, eval=FALSE}
library(CircaPower)

## Calculate power given (1) sample size n, (2) intrinsic effect size r, and (3) alpha level
CircaPower(n=12, r=1.5, alpha = 0.05)

## Calculate sample size given (1) pre-specified power, (2) intrinsic effect size r, and (3) alpha level
CircaPower(power=0.8, r=1.5, alpha = 0.05)

```
  
## 2. Circadian power calculation (passive design) given irregular circadian time distribuiton.

For irregular circadian time distribuiton, the sampling design factor d is dependent on the phase value $\phi$ which needs to be pre-specified. 

When the number of given irregular circadian times is different from the desired sample size, our package provides two estimation approaches for estimating the sampling design factor d in power calculation: 

(1) ct_estimation="expected" where d is calculated from the cts provided directly; 
(2) ct_estimation = "sampling" where n circadian times are first draw from the kernel density estimated from the empirical distribution of the cts provided and the n circadian times will be used for caclculating d.


```{r, eval=FALSE}
library(CircaPower)

## Use observed time of death from the human post-mortem brain transcriptomic study on Brodmann's area 11 and 47 in the prefrontal cortex as an example.
data(cts_Chen)

## Calculate power given (1) sample size n, (2) intrinsic effect size r, (3) phase value phi, (4) irregular circadian times cts and (5) alpha level using cts provided for estimating d.
CircaPower(n=100, r=0.4, phi=0, cts=cts_Chen, ct_estimation="expected", alpha = 0.05)

## Calculate power given (1) sample size n, (2) intrinsic effect size r, (3) phase value phi, (4) irregular circadian times cts and (5) alpha level using sampling approach for estimating d.
CircaPower(n=100, r=0.4, phi=0, cts=cts_Chen, ct_estimation="sampling", alpha = 0.05)

```
  
For the inverse problem of deriving sample size given pre-specified power, the "expected" approach is used for estimating d.
```{r, eval=FALSE}
## Calculate sample size given (1) pre-specified power, (2) intrinsic effect size r, (3) phase value phi, (4) irregular circadian times cts and (3) alpha level.
CircaPower(power=0.75, r=0.4, phi=0, cts=cts_Chen, alpha = 0.05)

```
## 3. A case study how to perform circadian power calculation using mouse skeletal muscle pilot data.
We first estimate the intrinsic effect sizes from this pilot data for: (i) median r of the 7 core circadian genes; (ii) minimum r of the top 100 significant circadian genes. The resulting intrinsic effect sizes are 3.58 and 2.23 respectively. Then, we can calculate the power at any desired sample size by inserting these estimated r.

```{r, eval=FALSE}
## Calculate power using the median r of the 7 core circadian genes
CircaPower(n=12, r=3.58, alpha = 0.05)

## Calculate power using the minimum r of the top 100 significant circadian genes
CircaPower(n=12, r=2.23, alpha = 0.05)

```

## 4. Intrinsic effect size from public databases.
We obtain the estimated intrinsic effect sizes from 3 publicly available human (@chen2016effects, @ketchesin2021diurnal and @seney2019diurnal) and 14 mouse (@aguilar2013cycles, @bray2008disruption, @cho2012regulation, @gerstner2016removal, @hoogerwerf2008transcriptional, @hughes2009harmonics, @mari2016gut, @masri2016lung, @masri2014partitioning, @na2009comprehensive, @nikolaeva2012circadian, @paschos2012obesity, @solanas2017aged, @zhang2014circadian) transcriptomic circadian studies. When a user needs to perform power calculation without pilot data, information from these public data can be used as a reference resource to facilitate circadian power calculation.


```{r ,echo = FALSE,out.width='100%',fig.cap="Table1: Intrinsic effect sizes for public available transcriptomic circadian data, including 3 passively designed human postmortem brain studies and 14 actively designed mouse studies from 20 types of tissues. These data are processed using the cosinor method (Cornelissen (2014)). Two types intrinsic effect sizes are used: (i) median $r$ of the 7 core circadian genes; (ii) minimum $r$ of the top 100 significant circadian genes."}
knitr::include_graphics("effectSizeTable.png")
```

***
# References